hasher-matcher-actioner/hmalib/fetching/tests/test_fetcher_and_bank_store.py

```diff
@@ -0,0 +1,310 @@
+threatexchange.exchanges.tests.test_state
+
+Verifies (using a mock bank) that the operations issued on a bank when fetching
+a collab are what we'd expect.
+"""
+
+import typing as t
+import contextlib
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import datetime
+import pytest
+import boto3
+from moto import mock_dynamodb2
+
+from threatexchange.signal_type.signal_base import SignalType
+from threatexchange.exchanges.signal_exchange_api import TSignalExchangeAPICls
+from threatexchange.exchanges.collab_config import (
+    CollaborationConfigBase,
+    CollaborationConfigWithDefaults,
+)
+from threatexchange.utils.dataclass_json import dataclass_dumps
+from threatexchange.exchanges.signal_exchange_api import SignalExchangeAPI
+from threatexchange.exchanges.fetch_state import (
+    FetchCheckpointBase,
+    TUpdateRecordKey,
+    TUpdateRecordValue,
+    SignalOpinion,
+    SignalOpinionCategory,
+    FetchedSignalMetadata,
+    FetchDelta,
+)
+from threatexchange.signal_type.md5 import VideoMD5Signal
+from threatexchange.signal_type.pdq.signal import PdqSignal
+from threatexchange.content_type.text import TextContent
+from threatexchange.content_type.content_base import ContentType
+
+from hmalib.fetching.fetcher import Fetcher
+from hmalib.fetching.bank_store import BankCollabFetchStore
+from hmalib.common.mappings import HMASignalTypeMapping
+from hmalib.common.models.bank import (
+    BanksTable,
+    Bank,
+    BankMember,
+    BankMemberSignal,
+    PaginatedResponse,
+)
+from hmalib.common.config import HMAConfig, create_config
+from hmalib.common.configs.tx_collab_config import EditableCollaborationConfig
+from hmalib.common.configs.tx_apis import ToggleableSignalExchangeAPIConfig
+
+from hmalib.common.tests.mapping_common import get_default_signal_type_mapping
+from hmalib.common.models.tests.ddb_test_common import (
+    DynamoDBTableTestBase,
+    HMAConfigTestBase,
+)
+from hmalib.common.models.tests.test_signal_uniqueness import BanksTableTestBase
+
+
+@dataclass
+class FakeCheckpoint(FetchCheckpointBase):
+    update_time: int
+
+
+@dataclass
+class FakeUpdateRecord:
+    # Interpret all values None as delete
+    signal_type: t.Optional[t.Type[SignalType]]
+    signal_value: t.Optional[str]
+    tag: t.Optional[str]
+
+
+@dataclass
+class FakeSignalMetadata(FetchedSignalMetadata):
+    tags: t.Set[str]
+
+    def get_as_opinions(self) -> t.Sequence[SignalOpinion]:
+        return [
+            SignalOpinion(True, SignalOpinionCategory.INVESTIGATION_SEED, self.tags)
+        ]
+
+
+# terrible-design, but it is a test. module level state to be overriden before
+# creating a fetcher instance. Since FakeAPI is instantiated via a classmethod
+# and FakeAPI needs to be discoverable as hmalib.foo.bar.tests.FakeAPI, I can't
+# use any kind of dynamic class gen.
+FETCH_RESPONSES: t.Sequence[
+    t.Dict[TUpdateRecordKey, t.Optional[TUpdateRecordValue]]
+] = []
```

Comment:
The secret here is to move this into a dedicated python fixture, or to use monkeypatch (a py.test fixture equivalent to mock.patch)